toolchain/check/deduce.cpp

```diff
@@ -246,54 +270,68 @@ auto DeductionContext::Deduce() -> bool {
-              CARBON_DIAGNOSTIC(
-                  InitializingGenericParam, Note,
-                  "initializing generic parameter `{0}` declared here",
-                  SemIR::NameId);
-              builder.Note(
-                  param_id, InitializingGenericParam,
-                  context().entity_names().Get(param->entity_name_id).name_id);
-            }
-          });
+      // TODO: Only do this when deducing against a symbolic pattern.
+      DiagnosticAnnotationScope annotate_diagnostics(&context().emitter(),
+                                                     note_initializing_param);
       arg_id = ConvertToValueOfType(context(), loc_id_, arg_id, param_type_id);
       if (arg_id == SemIR::InstId::BuiltinError) {
         return false;
       }
     }
 
-    // If the parameter is a symbolic constant, deduce against it. Otherwise, we
-    // assume there is nothing to deduce.
-    // TODO: This won't do the right thing in a template deduction.
-    auto param_const_id = context().constant_values().Get(param_id);
-    if (!param_const_id.is_valid() || !param_const_id.is_symbolic()) {
-      continue;
-    }
-
     // Attempt to match `param_inst` against `arg_id`. If the match succeeds,
     // this should `continue` the outer loop. On `break`, we will try to desugar
     // the parameter to continue looking for a match.
-    auto param_inst = context().insts().Get(
-        context().constant_values().GetInstId(param_const_id));
+    auto param_inst = context().insts().Get(param_id);
     CARBON_KIND_SWITCH(param_inst) {
-      // Deducing a symbolic binding from an argument with a constant value
-      // deduces the binding as having that constant value.
-      case SemIR::InstKind::SymbolicBindingPattern:
-      case SemIR::InstKind::BindSymbolicName: {
-        auto entity_name_id = SemIR::EntityNameId::Invalid;
-        if (auto bind = param_inst.TryAs<SemIR::SymbolicBindingPattern>()) {
-          entity_name_id = bind->entity_name_id;
-        } else {
-          entity_name_id =
-              param_inst.As<SemIR::BindSymbolicName>().entity_name_id;
+      // Deducing a symbolic binding pattern from an argument deduces the
+      // binding as having that constant value. For example, deducing
+      // `(T:! type)` against `(i32)` deduces `T` to be `i32`. The argument is
+      // required to be a compile-time constant.
+      case CARBON_KIND(SemIR::SymbolicBindingPattern bind): {
+        auto& entity_name = context().entity_names().Get(bind.entity_name_id);
+        auto index = entity_name.bind_index;
+        if (!index.is_valid()) {
+          break;
+        }
+        CARBON_CHECK(
+            index >= first_deduced_index_ &&
+                static_cast<size_t>(index.index) < result_arg_ids_.size(),
+            "Unexpected index {0} for symbolic binding pattern; "
+            "expected to be in range [{1}, {2})",
+            first_deduced_index_.index, result_arg_ids_.size());
```

Comment:
It looks like the format string wants 3 arguments, but I only see 2.

Reply:
Thanks, fixed, and tested manually by forcing the CHECK to fail.